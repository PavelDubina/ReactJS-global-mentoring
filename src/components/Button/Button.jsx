import React from 'react'
import PropTypes from 'prop-types'
import styles from './Button.scss'

export const Button = ({ isClose, styleType, children, onClick }) => (
  <button data-close={isClose} onClick={onClick} className={styles[styleType]}>
    {children}
  </button>
)

Button.propTypes = {
  isClose: PropTypes.bool,
  children: PropTypes.node.isRequired,
  styleType: PropTypes.string.isRequired,
  onClick: PropTypes.func,
}
