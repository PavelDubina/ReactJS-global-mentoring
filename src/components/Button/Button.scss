button {
  width: 144px;
  font-size: 18px;
  border: none;
  border-radius: 4px;
  letter-spacing: 1px;
  outline: none;
  cursor: pointer;

  &:active {
    opacity: 0.95;
  }
}

.adding {
  padding: 11px 0;
  background-color: #55555585;
  color: #f65261;
}

.search {
  width: 198px;
  padding: 16px 0;
  background-color: #f65261;
  color: #fff;
}

.confirm {
  padding: 12px 0;
  background-color: #f65261;
  color: #fff;
}

.reset {
  padding: 11px 0;
  background-color: transparent;
  border: 1px solid #f65261;
  color: #f65261;
}

.disabled {
  padding: 11px 0;
  background-color: #55555585;
  border: 1px solid #fff;
  cursor: auto;
}
