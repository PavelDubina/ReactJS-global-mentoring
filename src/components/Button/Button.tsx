import React from 'react'
import styles from '../../../styles/Button.module.scss'

export enum EType {
  button = 'button',
  submit = 'submit',
  reset = 'reset',
}

export enum EStyleTypeBtn {
  adding = 'adding',
  search = 'search',
  confirm = 'confirm',
  reset = 'reset',
  disabled = 'disabled',
  back_home = 'back_home',
}

export type ButtonProps = {
  type?: EType
  styleType: EStyleTypeBtn
  children: React.ReactNode
  onClick?: () => void
}

// PATTERN:{Destructuring props}
export const Button: React.FC<ButtonProps> = React.memo(({ type = EType.button, styleType, children, onClick }) => (
  <button type={type} onClick={onClick} className={styles[styleType]}>
    {children}
  </button>
))
