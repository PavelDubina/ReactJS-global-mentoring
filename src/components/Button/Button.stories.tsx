import React from 'react'
import { Story, Meta } from '@storybook/react'
import { Button, ButtonProps, EStyleTypeBtn, EType } from './Button'

export default {
  title: 'MyStorybook/Buttons',
  component: Button,
} as Meta

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Default = Template.bind({})
Default.args = {
  styleType: EStyleTypeBtn.adding,
  children: '+ ADD MOVIE',
  type: EType.button,
}
