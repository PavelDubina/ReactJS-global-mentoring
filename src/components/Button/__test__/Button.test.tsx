import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Button, EStyleTypeBtn } from '../Button'

describe('Button', () => {
  const getComponent = ({ onClick = jest.fn(), styleType = EStyleTypeBtn.adding }) =>
    render(
      <Button onClick={onClick} styleType={styleType}>
        Test Button
      </Button>,
    )
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('onClick have been called', () => {
    const click = jest.fn()
    const { getByText } = getComponent({ onClick: click })
    expect(getByText('Test Button')).toBeTruthy()
    fireEvent.click(getByText('Test Button'))
    expect(click).toBeCalled()
  })
})
