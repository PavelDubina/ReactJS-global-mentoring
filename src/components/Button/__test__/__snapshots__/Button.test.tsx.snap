// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should renders correctly 1`] = `
<DocumentFragment>
  <button
    class="adding"
    type="button"
  >
    Test Button
  </button>
</DocumentFragment>
`;
